Add bloganuary banner to reader

roo2 opened the pull request:
Adds a banner to the reader p1701190496797509-slack-C03NLNTPZ2T.

I've added it to the /read and /tag/$tag pages, we could add it to more pages if needed.


<img width="715" alt="Screen Shot 2023-12-11 at 11 49 50 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/23f4e71a-f63c-475e-b989-e165c940ff34">

### Testing instructions 
The banner is only active if $bloganuary is true, test with:
`/read?flags=+bloganuary`
`/tag/art?flags=+bloganuary`

roo2 committed:
add bloganuary header to reader

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,39 @@
+	const translate = useTranslate();
+
+	const trackBloganuaryMoreInfoClick = () => {
+		dispatch( recordTracksEvent( 'reader_bloganuary_more_info_click' ) );
```
We should probably change the name to fit with the other event standards with a `calypso` prefix. ex. `calypso_reader_bloganuary_banner_more_info_click`

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,39 @@
+	const trackBloganuaryMoreInfoClick = () => {
+		dispatch( recordTracksEvent( 'reader_bloganuary_more_info_click' ) );
+	};
+
```
We could benefit from a `viewed` type event to be able to understand user interaction banner. ex. `calypso_reader_bloganuary_banner_viewed` being sent on mount.

One option is putting this in a useEffect on mount and also checking the `isBloganuary` condition to see if we should fire the event. Although it feels a little weird checking conditions that represent it rendering the component vs. it actually rendering the component. So another thought would be to use the useEffect on mount, but also add an outer wrapper component to this one so that it never renders outside of Bloganuary:


```
export default const BloganuaryHeaderWrapper = () => {
   if (! isBloganuary() ) return null;

  return <BloganuaryHeader />
}
```

And then we wouldn't have to rely on a check for `isBloganuary` in the `<BloganuaryHeader />`'s useEffect for determining whether or not to send the viewed event since render would imply viewed in that case.

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,39 @@
+				<span className="bloganuary-header__title">{ translate( 'Bloganuary' ) }</span>
+			</div>
+			<a
+				href="https://wordpress.com/bloganuary"
```
Do we need to localize this url?  ex `https://wordpress.com/${getLocale()}/bloganuary` or something similar?

roo2 replied:
ahh yes I think so

Addison-Stavlo reviewed (commented):
A couple notes and Qs:

roo2 committed:
subscribe to bloganuary tag from reader banner

roo2 commented:
Thanks for the reviews @Addison-Stavlo.

I've addressed the feedback and added a "Follow bloganuary tag" button after confirming with Mindy pe7F0s-1iI-p2#comment-1835.

I'm ~90% sure I'm using redux and hooks as intended to handle displaying the success message, but maybe there's a simpler way to do it, I still find it a bit strange that you can't just pass in an onSuccess callback 😂

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,90 @@
+
+	useEffect( () => {
+		dispatch( recordTracksEvent( 'calypso_reader_bloganuary_banner_view' ) );
+	}, [ dispatch ] );
```
It feels weird using `dispatch` as a dependency in these cases (although I get the linters highlight it when its used in the hook). Ideally, dispatch should never change, but if for some reason its reference in memory changes this would result in sending the `*view` tracking event extra times and messing with the counts past what an empty dependency array would. 🤔 

I think we have other methods of `recordTracksEvent` that don't need to be used with dispatch, the dispatch version builds on them, but im unsure if its actually buying us anything valuable in this case.  I asked about this in slack a few weeks back p1697648257681619-slack-C03N25JPCE4

roo2 replied:
good catch! ahhh yes, I thought I'd seen recordTracksEvent work without dispatch! it was a different function with the same name 😆

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,90 @@
+
+	// Handle successfully subscribing to 'bloganuary' the redux-y way
+	// Errors are handled in client/state/data-layer/wpcom/read/tags/mine/new/index.js:receiveError
+	useEffect( () => {
+		// tripple equals check is needed because 'undefined' means the value has not been loaded yet.
+		if ( prevIsFollowingBloganuary.current === false && isFollowingBloganuary ) {
+			dispatch( successNotice( translate( 'Subscribed to bloganuary tag' ) ) );
+		}
+		prevIsFollowingBloganuary.current = isFollowingBloganuary;
+	}, [ isFollowingBloganuary, prevIsFollowingBloganuary, dispatch, translate ] );
```
This does seem a bit funky to handle the success notice. I think ideally the success notice should be dispatched on the redux side when the request to follow the tag receives a valid response: that way its not implemented in each component we use the action from, but tied to the action itself.

I don't see success messages for when we follow tags from the current follow tags buttons in the tags streams or on the discover tags tabs sidebar (at least not on production or calypso.live). So I don't think we need this here specifically. But if we did, we should probably fix it for all of the follow tags buttons via the action on the redux side as mentioned above. That seems out of scope for this PR though.

roo2 replied:
ohhh yes, you're right, I added this before I changed the button state, I originally thought we need *some* way to know that it was successful. But I think you're right, everywhere else we just change the button state.

Addison-Stavlo commented on client/components/bloganuary-header/index.tsx:
```diff
@@ -0,0 +1,90 @@
+					onClick={ subscribeToBloganuaryTag }
+					disabled={ isFollowingBloganuary }
+				>
+					{ isFollowingBloganuary ? translate( 'Subscribed' ) : translate( 'Subscribe' ) }
```
I think we still use the term `Follow` (or `Follow tag`) for tags, and `Subscribe` for sites. Followed tags aren't something that can be found in "Manage subscriptions" settings, so it may be weird referring to them that way.

roo2 replied:
ahhhh ok, thanks for clarifying

roo2 committed:
add unfollow and remove noytification

roo2 committed:
use follow button component

roo2 commented:
I addressed feedback and updated to use the follow button component which can follow and unfollow the tag.

eoigal reviewed (approved):
This worked for me but I got a different header to the one in the example in PR description

<img width="1055" alt="Screenshot 2023-12-14 at 17 36 30" src="https://github.com/Automattic/wp-calypso/assets/5560595/824aeb7c-4836-49b3-865b-c131651fbbfe">

I wasn't sure about the Follow button in the header, was it a button to follow the bloganuary or the tag I was looking at... I assumed the former. Not sure how that can be clearer though.

Other than that, everything worked as I'd expect.

One question - are we going to add the 'Bloganaury' tag to the discover top tag menu?

Addison-Stavlo reviewed (approved):
WFM!

> One question - are we going to add the 'Bloganaury' tag to the discover top tag menu?

I don't think this was on the radar nor had I thought of it.  We could probably replace the dailyprompt tab there with bloganuary when bloganuary is active fairly easily. Not in this PR though.

roo2 commented:
thanks @eoigal I also saw from your screenshot that the icon alignment was off in FF 😆. 

> I wasn't sure about the Follow button in the header, was it a button to follow the bloganuary or the tag I was looking at... I assumed the former. Not sure how that can be clearer though.

Yea I can't see a way to clarify, if I made it say "Follow 'bloganuary' tag", that would take up most of the banner, I think it's ok if people aren't completely sure, worst case they might try it and see that bloganuary gets added to the sidebar and the tag doesn't get followed 🤷‍♂️

roo2 committed:
disable follow button for logged out and fix icon alignment in FF

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10331400

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @roo2 for including a screenshot in the description! This is really helpful for our translators.

Addison-Stavlo commented:
> One question - are we going to add the 'Bloganaury' tag to the discover top tag menu?

I opened a PR for this here - https://github.com/Automattic/wp-calypso/pull/85429 - feel free to pick it up while I am afk if it is not yet merged.  cc @roo2

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10331400).
